### DISCUSSION TITLE
Setting up the controller for MPE

### DISCUSSION THREAD

**Discussion by Ferdzz** [FortySevenEffects/arduino_midi_library#187] [Q&A]
Hi, I'm really glad to see that you've enabled Github Discussions!

I'm currently working on a breath-powered MIDI controller, and for it to work properly, I need to use the MPE (Midi Polyphonic Expression) mode. From the documentation on [midi.org](https://www.midi.org/specifications/midi1-specifications/mpe-midi-polyphonic-expression), I read that it may be necessary to signal my controller as an MPE capable board. 

How can I send these "Zones" and "Modes" definitions using your library? I'm new to implementing Midi in my projects, so please tell me if I'm on the wrong track here.

I was able to send expressions CCs, notes, and etc. Even when splitting both of my notes and CC 11 (Expression) on separate channels, Garageband appears to apply the CC on both notes. Could I get some help with setting up for proper MPE communications? 

Thank you very much

**Comment by franky47** [comment]
I'm not familiar with MPE, let me read the spec and I'll get back to you.

**Comment by BenjaminPoilve** [comment]
Just to add to this: I've tried the same and here where I stand: 
- For most uses, MIDI mode 4 seems to be enough. MPE is just a standardisation layer over it. 
- It is quite easy to have one note per channel and a per-channel bend signal. 
- Where I am personally stuck is on the per-channel volume control. According to specs, the CC#7 should now be a channel volume. The issue is that all softsynth that I tried seems to apply it to all channels. It is not the case for other CC (for example PAN is correctly interpreted as per-channel). 

I have made a short demo sketch that takes three notes, slides one up and one down (I'm on a Raspberry 2040 chip):
```
#include "Arduino.h"
#include "Adafruit_TinyUSB.h"
#include "MIDI.h"

//>>>>MIDI CHARACTERISTICS - USB MIDI
Adafruit_USBD_MIDI usb_midi;
MIDI_CREATE_INSTANCE(Adafruit_USBD_MIDI, usb_midi, MIDI_USB);

int bend_multiplier[]={0,1,-1};
int channels[]={2,3,4};
int volumes[]={127,0,0};
int volume_control_CC=7;

void setup() {
  // Manual begin() is required on core without built-in support e.g. mbed rp2040
  if (!TinyUSBDevice.isInitialized()) {
    TinyUSBDevice.begin(0);
  }

  Serial.begin(115200);

  // Initialize MIDI, and listen to MIDI channel 1
  delay(500);
  MIDI_USB.begin(1);
  delay(500);
  // If already enumerated, additional class driverr begin() e.g msc, hid, midi won't take effect until re-enumeration
  if (TinyUSBDevice.mounted()) {
    TinyUSBDevice.detach();
    delay(10);
    TinyUSBDevice.attach();
  }
}

void loop() {
  #ifdef TINYUSB_NEED_POLLING_TASK
  // Manual call tud_task since it isn't called by Core's background
  TinyUSBDevice.task();
  #endif
  MIDI_USB.read();

  //setting the mode of the synthetiser (mode 4), goes on base channel of the device
  for(int i=1;i<16;i++){ //not sure which one is the control channel 
    MIDI_USB.sendControlChange(124, 0,i); //OMNI OFF
    MIDI_USB.sendControlChange(126, 3,i); //MONO 3 VOICES
    //setting the noterange of the synthetiser
    MIDI_USB.sendControlChange(101, 0, i);
    MIDI_USB.sendControlChange(100, 0,i);
    MIDI_USB.sendControlChange(6, 12,i);
    MIDI_USB.sendControlChange(101, 127, i);
    MIDI_USB.sendControlChange(100, 127,i);
    delay(5);
  }
  for(int index=0;index<3;index++){
    MIDI_USB.sendNoteOn(60, 127, channels[index]);   
    MIDI_USB.sendPitchBend(0 ,channels[index]);   
    MIDI_USB.sendControlChange(volume_control_CC, volumes[index],channels[index]);
  }
  for(int step=0;step<4000;step+=10){
    for(int index=0;index<3;index++){
      MIDI_USB.sendPitchBend(bend_multiplier[index]*step,channels[index]);
      MIDI_USB.sendControlChange(volume_control_CC, volumes[index], channels[index]);
    }
    delay(10);
  }
  for(int index=0;index<3;index++){
    MIDI_USB.sendNoteOff(60, 127, channels[index]);     
  }
  delay(1000);
}
```

When all volumes are on the same level, it works, but if they are different it gets all distorted as the volumes are successively applied to all channels. 

Checking the signal with MidiView, nothing seems out of place, all signals are associated to their proper channels. 

I am missing something obvious?
